src/database/utilities/createRecord.js

```diff
@@ -82,63 +130,134 @@ const createAddress = (database, { line1, line2, line3, line4, zipCode } = {}) =
+const getPatientUniqueCode = database => {
   const { PATIENT_CODE } = NUMBER_SEQUENCE_KEYS;
-  const { dateOfBirth, addressOne, addressTwo, lastName, firstName } = patientDetails;
-
-  const billingAddress = createAddress(database, { line1: addressOne, line2: addressTwo });
-
-  const thisStoreId = database.getSetting(SETTINGS_KEYS.THIS_STORE_ID);
-  const thisStoreCode = database.getSetting(SETTINGS_KEYS.THIS_STORE_CODE);
   const patientSequenceNumber = getNextNumber(database, PATIENT_CODE);
-  const uniqueCode = `${thisStoreCode}${String(patientSequenceNumber)}`;
+  const thisStoreCode = database.getSetting(SETTINGS_KEYS.THIS_STORE_CODE);
+  return `${thisStoreCode}${String(patientSequenceNumber)}`;
+}
 
-  const fullName = `${lastName}, ${firstName}`;
+/**
+ * Creates a new patient record. Patient details passed can be in the shape:
+ *  {
+ *    id, name, firstName, lastName, code, dateOfBirth, phoneNumber, emailAddress,
+ *    billAddress1, billAddress2, billAddress3, billAddress4, billPostalZipCode,
+ *    country, supplyingStoreId, 
+ *  }
+ */
+const createPatient = (database, patientDetails) => {
+  const {
+    id: patientId,
+    name: patientName,
+    firstName,
+    lastName,
+    code: patientCode,
+    dateOfBirth: patientDateOfBirth,
+    phoneNumber,
+    emailAddress,
+    billAddress1, 
+    billAddress2, 
+    billAddress3,
+    billAddress4, 
+    billPostalZipCode,
+    country,
+    supplyingStoreId: patientSupplyingStoreId,
+  } = patientDetails;
+
+  const id = patientId ?? generateUUID();
+  const name = patientName ?? `${lastName}, ${firstName}`;
+  const code = patientCode ?? getPatientUniqueCode(database);
+  const dateOfBirth = patientDateOfBirth ?? new Date();
+
+  const type = 'patient';
+  const isPatient = true;
+  const isCustomer = true;
+  const isSupplier = false;
+  const isManufacturer = false;
+
+  const billingAddress = getOrCreateAddress(
+    database,
+    billAddress1,
+    billAddress2,
+    billAddress3,
+    billAddress4,
+    billPostalZipCode
+  );
 
-  const patient = database.create('Name', {
-    id: generateUUID(),
-    ...patientDetails,
-    billingAddress,
+  const thisStoreId = database.getSetting(SETTINGS_KEYS.THIS_STORE_ID)
+  const supplyingStoreId = patientSupplyingStoreId ?? thisStoreId;
+  const thisStoresPatient = supplyingStoreId === thisStoreId;
+  const isVisible = true;
+  
+  const patient = database.update('Name', {
+    id,
+    name,
+    code,
+    type,
     dateOfBirth,
-    isVisible: true,
-    isPatient: true,
-    type: 'patient',
-    code: uniqueCode,
-    supplyingStoreId: thisStoreId,
-    isCustomer: true,
-    thisStoresPatient: true,
-    name: fullName,
+    isPatient,
+    isCustomer,
+    isSupplier,
+    isManufacturer,
+    phoneNumber,
+    emailAddress,
+    billingAddress,
+    country,
+    supplyingStoreId,
+    thisStoresPatient,
+    isVisible,
   });
 
-  database.save('Patient', patient);
+  database.save('Name', patient);
```

Comment:
don't need to `save` if you use update

Reply:
Nice, out of curiousity, what is the need for `create`? After referencing with `react-native-database`, `update` seems to cover all `create` use-cases (seems to be equivalent of an `upsert`, so creates if primary key non-existent but updates if the record already exists).

Reply:
3f1ef95d0f48b6cdf328930a001b6f50b187ecb1.